NiftyReg dot() type error

for something like: 

```
examples_path = SIRF_PATH + '/data/examples/Registration'
x = sirf.Reg.NiftiImageData3D(rexamples_path + "/test.nii.gz") #ref_aladin 
T = na.get_transformation_matrix_forward()
y = sirf.Reg.NiftiImageData3D(examples_path + "/test2.nii.gz") #flo_aladin

nr = sirf.Reg.NiftyResample()
nr.set_reference_image(x)
nr.set_floating_image(y)
nr.set_interpolation_type_to_linear()
nr.add_transformation(T)
# Add this on test_pReg.py#L950 to test.

# NOTE: you need PR #626 for this to work. Otherwise shaw domain/range
x = nr.domain_geometry().allocate(value = 'random')
y = nr.range_geometry().allocate( value = 'random')
y_hat = nr.direct(x)
this_errors = y_hat.dot(y)
```
The error reads:

```
File "/home/anderbiguri/SIRF/build/sources/SIRF/src/Registration/pReg/tests/test_pReg.py", line 960, in try_niftymomo
this_errors = y_hat.dot(y)
File "/home/anderbiguri/SIRF/build/INSTALL/python/sirf/SIRF.py", line 101, in dot
assert_validities(self, other)
File "/home/anderbiguri/SIRF/build/INSTALL/python/sirf/Utilities.py", line 396, in assert_validities      type(y)))
AssertionError: Expecting same type input, got <class 'sirf.Reg.NiftiImageData'> and <class 'sirf.Reg.NiftiImageData3D'>
```

Apparently it returns a different class when you do `allocate()` (`NiftiImageData3D`, as the original variable type) and when you do `direct()` (`NiftiImageData`), but both are  supposed to be conceptually the same thing.

Any suggestions on how to fix this are welcomed. 

I can think of 

1. Change `direct()` to return `NiftiImageData3D`. This will likely cause lots of errors down the chain of functions that expect this method's output to be `NiftiImageData`. 
2. Change `dot()` so it allows some cases of different classes as input. 


You can test this code adding it on `test_pReg.py#L950` and using PR #626 (or simply swap domain/range in the code).

If you look at the source code (`Reg.py`), does the `.dot` method check that the argument type matches that of `self`? From the error: `AssertionError: Expecting same type input`, I'm guessing that this is the case.

Probably just changing it to check that the input is of type `NiftiImageData` would solve this one.

--- a/src/Registration/pReg/Reg.py.in
+++ b/src/Registration/pReg/Reg.py.in
@@ -237,6 +237,11 @@ class NiftiImageData(SIRF.ImageData):
         """Get sum."""
         return parms.float_par(self.handle, 'NiftiImageData', 'sum')
 
+    def dimensions(self):
+        """Returns image dimensions as a tuple (nz, ny, nx)."""
+        dim  = self.get_dimensions()
+        return tuple(dim[1:1+dim[0]]) # dim[0] tells us how many dimensions
+
     def get_dimensions(self):
         """Get dimensions. Returns nifti format.
         i.e., dim[0]=ndims, dim[1]=nx, dim[2]=ny,..."""

--- a/src/common/SIRF.py
+++ b/src/common/SIRF.py
@@ -506,11 +506,15 @@ class DataContainer(ABC):
         return self.__div__(other)
     @property
     def shape(self):
-        '''returns the shape of the data array
+        '''Returns the shape of the data array
         
         CIL/SIRF compatibility
         '''
         return self.as_array().shape
+    @property
+    def size(self):
+        '''Returns the (total) size of the data array.'''
+        return self.as_array().size
 
 class ImageData(DataContainer):
     '''

--- a/src/common/Utilities.py
+++ b/src/common/Utilities.py
@@ -390,7 +390,7 @@ def assert_validity(obj, dtype):
 
 def assert_validities(x, y):
     try:
-        assert type(x) == type(y)
+        assert issubclass(type(x),type(y)) or issubclass(type(y),type(x)) # returns true if both are the same class
     except AssertionError as ae:
         raise AssertionError('Expecting same type input, got {} and {}'.format(type(x), 
                                                                                type(y)))
@@ -398,6 +398,8 @@ def assert_validities(x, y):
         raise AssertionError('handle for first parameter is None')
     if y.handle is None:
         raise AssertionError('handle for second parameter is None')
+    if x.dimensions() != y.dimensions():
+        raise ValueError("Input shapes are expected to be equal, got " + str(x.dimensions()) + " and " + str(y.dimensions()) + " instead.")
 
 
 def label_and_name(g):

